### PR TITLE
feat: CommandEnvironment

### PR DESCRIPTION
**Pull request by Khaazz**
# PULL REQUEST

## **Overview**

This PR adds the `CommandEnvironment` structure to abstract and encapsulate the arguments being used in `Command.execute()`.  
This also simplify internals about resolving arguments and will bring an easy support for multiple ArgumentParser types in the future.

The `CommandEnvironment` object won't change anything about using `Command.execute()` but brings additional information and ease resolving the different variables needed.

## **Status**

- [X] Typings have been updated or don't need to be.
- [X] This PR have been tested and is ready to be merged.

## **Semantic versioning classification**

- [ ] This PR introduces *BREAKING* changes.
- [X] This PR adds new features, improve the code and implies minimal changes.
- [ ] This PR fixes a bug and references the relevant issue or documentation.
- [ ] This PR improve performance or code refactor without API changes.
- [ ] This PR **only** includes non-code changes (documentation, CI, tools...).

### CODE REVIEW

**Comment by Khaazz** [comment]
@bsian03 Please could you review jsdoc and the overall system?
Open to any suggestion.
This will need to be reflected in #59 as well.

**Comment by bsian03** [comment]
Looks good, I'll create reviews for type changes

File: examples/djs/src/Client.js [hunk 1]
```diff
@@ -52,9 +52,9 @@ class Client extends AxonClient {
 
     // disabled
     // eslint-disable-next-line no-unused-vars
-    $sendHelp(command, { msg, guildConfig, isAdmin, isOwner } ) {
+    $sendHelp(command, env, isAdmin, isOwner) {
```
Comment (bsian03) [inline examples/djs/src/Client.js]: 1

File: examples/djs/src/Client.js [hunk 2]
```diff
@@ -52,9 +52,9 @@ class Client extends AxonClient {
...
         // override sendHelp method
-        return this.axonUtils.sendMessage(msg.channel, `Help override for ${command.label}`);
+        return this.axonUtils.sendMessage(env.msg.channel, `Help override for ${command.label}`);
```
Comment (bsian03) [inline examples/djs/src/Client.js]: 1

File: examples/eris/src/Client.js [hunk 3]
```diff
@@ -49,9 +49,9 @@ class Client extends AxonClient {
 
     // disabled
     // eslint-disable-next-line no-unused-vars
-    $sendHelp(command, { msg, guildConfig, isAdmin, isOwner } ) {
+    $sendHelp(command, env, isAdmin, isOwner) {
```
Comment (bsian03) [inline examples/eris/src/Client.js]: 1

File: examples/eris/src/Client.js [hunk 4]
```diff
@@ -49,9 +49,9 @@ class Client extends AxonClient {
...
         // override sendHelp method
-        return this.axonUtils.sendMessage(msg.channel, `Help override for ${command.label}`);
+        return this.axonUtils.sendMessage(env.msg.channel, `Help override for ${command.label}`);
```
Comment (bsian03) [inline examples/eris/src/Client.js]: 1

File: src/AxonClient.js [hunk 5]
```diff
@@ -576,24 +577,19 @@ class AxonClient extends EventEmitter {
...
      * @param {Command} command
-     * @param {GuildConfig} guildConfig
-     * @param {Object} permissions
-     * @param {Boolean} permissions.isAdmin
-     * @param {Boolean} permissions.isOwner
+     * @param {CommandEnvironment} env
      */
-    _execCommand(msg, args, command, guildConfig, { isAdmin, isOwner } ) {
-        this.emit('debug', DEBUG_FLAGS.INFO | DEBUG_FLAGS.COMMAND, `${guildConfig ? '[GUILD]' : '[DM]'} ${isAdmin ? 'Admin' : 'Regular'} execution of ${command.fullLabel}`);
+    _execCommand(command, env) {
```
Comment (bsian03) [inline src/AxonClient.js]: 2

File: src/AxonClient.js [hunk 6]
```diff
@@ -620,29 +616,23 @@ class AxonClient extends EventEmitter {
...
-      * @param {Message} msg
-      * @param {Array<String>} args
-      * @param {Command} command
-      * @param {GuildConfig} guildConfig
-      * @param {Object} permissions
-      * @param {Boolean} permissions.isAdmin
-      * @param {Boolean} permissions.isOwner
-      */
-    _execHelp(msg, args, command, guildConfig, { isAdmin, isOwner } ) {
+     * @param {Command} command
```
Comment (bsian03) [inline src/AxonClient.js]: 2

File: src/AxonClient.js [hunk 7]
```diff
@@ -620,29 +616,23 @@ class AxonClient extends EventEmitter {
...
+        this.emit('debug', DEBUG_FLAGS.INFO | DEBUG_FLAGS.COMMAND, `${guildConfig ? '[GUILD]' : '[DM]'} ${env.isAdmin || env.isOwner ? 'Admin' : 'Regular'} -HELP- execution of ${command.fullLabel}`);
         let startTime;
         if (this.settings.debugMode) {
             startTime = performance.now();
         }
 
-        command.sendHelp( {
-            msg, args, guildConfig, isAdmin, isOwner,
-        } )
+        command.sendHelp(env)
```
Comment (bsian03) [inline src/AxonClient.js]: 2

File: src/Structures/Command/Command.js [hunk 8]
```diff
@@ -194,19 +194,14 @@ class Command extends Base {
...
-     * @param {Array<String>} params.args
-     * @param {GuildConfig} params.guildConfig
-     * @param {Boolean} params.isAdmin
-     * @param {Boolean} params.isOwner
+     * @param {CommandEnvironment} env
      * @returns {Promise<CommandContext>} Return a CommandContext or throw an AxonCommandError.
      * @memberof Command
      */
-    _process(params) {
+    _process(env) {
```
Comment (bsian03) [inline src/Structures/Command/Command.js]: 2

File: src/Structures/Command/Command.js [hunk 9]
```diff
@@ -224,13 +219,13 @@ class Command extends Base {
...
                     executed: false,
-                    executionType: CommandContext.getExecutionType(isAdmin, isOwner),
+                    executionType: env.executionType,
                     executionState: COMMAND_EXECUTION_STATE.INVALID_PERMISSIONS_BOT,
                 } ).resolveAsync();
             }
 
             /* Permissions checkers */
-            if (!isAdmin) {
+            if (!env.isAdmin) {
```
Comment (bsian03) [inline src/Structures/Command/Command.js]: 3

Comment (bsian03) [reply src/Structures/Command/Command.js]: wait no

File: src/Structures/Command/Command.js [hunk 10]
```diff
@@ -308,25 +301,18 @@ class Command extends Base {
...
-     * @param {Boolean} object.isOwner
-     * @returns {CommandContext}
+     * @param {CommandEnvironment} env
+     * @returns {Promise<CommandContext>}
      * @memberof Command
      */
-    _execute( {
-        msg, args, guildConfig, isAdmin, isOwner,
-    } ) {
+    _execute(env) {
```
Comment (bsian03) [inline src/Structures/Command/Command.js]: 1

File: src/Structures/Command/Command.js [hunk 11]
```diff
@@ -346,11 +332,7 @@ class Command extends Base {
      * Override this method in all Command child.
      * Main method - command logic being executed when the command is actually ran.
      *
-     * @param {Object} object - An Object with all arguments to use execute
-     * @param {Message} [object.message] - The message Object
-     * @param {Array<String>} [object.args] - The Array of arguments
-     * @param {GuildConfig} [object.guildConfig] - The guildConfig if it exists
-     *
+     * @param {CommandEnvironment} env - The Command Environment object with all variables needed for the Commandexecution
```
Comment (bsian03) [inline src/Structures/Command/Command.js]: 1

File: src/Structures/Command/Command.js [hunk 12]
```diff
@@ -368,18 +350,15 @@ class Command extends Base {
...
-     * @param {Message} object.msg
-     * @param {GuildConfig} object.guildConfig
-     * @param {Boolean} object.isAdmin
-     * @param {Boolean} object.isOwner
+     * @param {CommandEnvironment} env
      * @returns {Promise<CommandContext>}
      * @memberof Command
      */
-    sendHelp( { msg, guildConfig, isAdmin, isOwner } ) {
+    sendHelp(env) {
```
Comment (bsian03) [inline src/Structures/Command/Command.js]: 1

File: src/Structures/CommandDispatcher.js [hunk 13]
```diff
@@ -13,6 +16,7 @@
  *
  * @class CommandDispatcher
  *
+ * @prop {AxonClient} _axon
```
Comment (bsian03) [inline src/Structures/CommandDispatcher.js]: 3

File: src/Structures/CommandDispatcher.js [hunk 14]
```diff
@@ -143,24 +152,19 @@ class CommandDispatcher {
      * It uses the global admin and owner prefixes and checks for the BotStaff rank of the caller.
      *
      * @param {Message} msg
-     * @returns {{isAdmin: Boolean, isOwner: Boolean}} { isAdmin: Boolean, isOwner: Boolean }
+     * @returns {COMMAND_EXECUTION_TYPES} The execution type
```
Comment (bsian03) [inline src/Structures/CommandDispatcher.js]: 3

File: examples/eris/src/Client.js (hunk 3 above)
Comment (Khaazz) [reply examples/eris/src/Client.js]: updated this. This now only takes `(command, env)`

File: examples/djs/src/Client.js (hunk 1 above)
Comment (Khaazz) [reply examples/djs/src/Client.js]: same here, updated this `(command, env)`

**Comment by Khaazz** [comment]
This PR will be merged when all the changes are reflected to #59 and after approval if everything is still good to merge.

File: src/Structures/Command/CommandEnvironment.js [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ * @author KhaaZ
+ *
+ * @class CommandEnvironment
+ * @prop {String} raw - The raw message content
+ * @prop {Message} msg - The message object from the lib
+ * @prop {Array<String>} args - The array of arguments
+ * @prop {GuildConfig} guildConfig - The GuildConfig data-structure with all DB saved settings
+ * @prop {String} prefix - The prefix used for this command
+ * @prop {String} command - The full label of the command being
```
Comment (bsian03) [inline src/Structures/Command/CommandEnvironment.js]: typo? 
> the command being

File: src/Structures/Command/CommandEnvironment.js [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    setCommand(command) {
+        this.command = command.fullLabel;
+        return this;
+    }
+
+    /**
+     * Resolve the argument from the args string.
+     * Uses the custom parser given in params
+     *
+     * @param {Object} parser
```
Comment (bsian03) [inline src/Structures/Command/CommandEnvironment.js]: Example format?

Comment (Khaazz) [reply src/Structures/Command/CommandEnvironment.js]: This is a pre implementation for the future argument parsing system. So atm nothing. I will document that later when adding the actual feature

File: src/Structures/Command/CommandEnvironment.js (hunk 15 above)
Comment (Khaazz) [reply src/Structures/Command/CommandEnvironment.js]: `being executed`. Typo indeed

Comment (Khaazz) [reply src/Structures/Command/CommandEnvironment.js]: 0fb77427e0b9cdb28d5f466e66236204206cc937

**Pull request merged**